### PR TITLE
Use CommentRanges in backwards lexing

### PR DESCRIPTION
**Pull request by konstin**
## Summary

The tokenizer was split into a forward and a backwards tokenizer. The backwards tokenizer uses the same names as the forwards ones (e.g. `next_token`). The backwards tokenizer gets the comment ranges that we already built to skip comments.

### CODE REVIEW

**Review by MichaReiser — COMMENTED** [review]
I really like the idea. It will simplify the tokenizer a lot. 

I think I would overall prefer to split the backward lexing out of the simple tokenizer because the changes now complicate the use of `SimpleTokenizer` even when backward lexing isn't necessary. This will also allow us to support fewer tokens for backward lexing.

File: crates/ruff_python_trivia/src/tokenizer.rs [hunk 1]
```diff
@@ -470,6 +470,174 @@ impl<'a> SimpleTokenizer<'a> {
...
+        // )
+        // ```
+        loop {
+            self.cursor.eat_back_while(is_python_whitespace);
+            if self.cursor.eat_char_back('\n') || self.cursor.eat_char_back('\r') {
+                self.cursor.eat_back_while(|c| matches!(c, '\n' | '\r'));
+                // Skip comments. This is easy in python since comments always end at the end of
+                // the line.
+                if let Ok(comment_idx) = comment_ranges
+                    .binary_search_by(|range| range.end().cmp(&self.cursor.text_len()))
```
[the remark this answers is not in the document]
Comment (konstin) [reply crates/ruff_python_trivia/src/tokenizer.rs]: i've implemented this but it feels like i added more complexity than appropriate for that small perf :/

File: crates/ruff_python_trivia/src/snapshots/ruff_python_trivia__tokenizer__tests__tokenize_bogus.snap [hunk 2]
```diff
@@ -21,106 +21,6 @@ expression: test_case.tokens()
...
-        range: 50..51,
-    },
-    SimpleToken {
-        kind: Bogus,
-        range: 51..52,
-    },
-    SimpleToken {
-        kind: Bogus,
-        range: 52..53,
+        range: 27..53,
```
Comment (MichaReiser) [inline crates/ruff_python_trivia/src/snapshots/ruff_python_trivia__tokenizer__tests__tokenize_bogus.snap:24]: Nit: Would you mind moving this change out of this PR and make a separate PR for it? It eases reviewing and gives us higher confidence that the behavior doesn't change with your refactor.

File: crates/ruff_python_trivia/src/tokenizer.rs [hunk 3]
```diff
@@ -762,322 +856,133 @@ impl<'a> SimpleTokenizer<'a> {
...
-                        }
-
-                        _ => {
-                            self.bogus = true;
-                            SimpleTokenKind::Other
-                        }
-                    }
-                }
-            }
+            '#' => unreachable!(),
```
Comment (MichaReiser) [inline crates/ruff_python_trivia/src/tokenizer.rs:868]: This could potentially happen if we start lexing before the `expr` in 

```python
f"#{expr}"
```

Let's return `Other` in this case.

File: crates/ruff_python_trivia/src/tokenizer.rs [hunk 4]
```diff
@@ -1167,9 +1024,14 @@ mod tests {
...
-            SimpleTokenizer::new(self.source, self.range)
-                .rev()
-                .collect()
+            let mut comment_ranges = Vec::new();
+            for result in lex(self.source, Mode::Module) {
+                let (token, range) = result.expect("Input to be a valid python program.");
+                if matches!(token, Tok::Comment(_)) {
+                    comment_ranges.push(range);
+                }
+            }
```
Comment (MichaReiser) [inline crates/ruff_python_trivia/src/tokenizer.rs:1033]: Nit
```suggestion
            let comment_ranges: Vec<_> = lex(self.source, Mode::Module).filter_map(|result| {
	             let (token, range) = result.expect("Input to be a valid python program.");
                if matches!(token, Tok::Comment(_)) {
                    Some(range)
                } else {
                	None
              	}
            }).collect()
```

File: crates/ruff_python_formatter/src/comments/mod.rs [hunk 5]
```diff
@@ -306,15 +306,31 @@ pub(crate) struct Comments<'a> {
...
     }
 
+    /// Effectively a [`Default`] implementation that works around the lifetimes for tests
+    #[cfg(test)]
+    pub(crate) fn from_ranges(comment_ranges: &'a CommentRanges) -> Self {
+        Self {
+            data: Rc::new(CommentsData::default()),
+            comment_ranges,
+        }
+    }
```
[the remark this answers is not in the document]
Comment (konstin) [reply crates/ruff_python_formatter/src/comments/mod.rs:328]: that doesn't work unfortunately because we store only a reference an the reference needs to outlive this function call

File: crates/ruff_python_trivia/src/tokenizer.rs [hunk 6]
```diff
@@ -523,23 +519,41 @@ impl<'a> SimpleTokenizer<'a> {
...
+            // Emit a single final bogus token
             let token = SimpleToken {
                 kind: SimpleTokenKind::Bogus,
-                range: TextRange::at(self.offset, first.text_len()),
+                range: TextRange::new(self.offset, self.source.text_len()),
             };
 
-            self.offset += first.text_len();
+            // Set the cursor to EOF
+            self.cursor.eat_while(|_| true);
```
[the remark this answers is not in the document]
Comment (konstin) [reply crates/ruff_python_trivia/src/tokenizer.rs:529]: is that fine? i wasn't sure if i could do that given that now the cursor and the tokenizer disagree about source

File: crates/ruff_python_trivia/src/tokenizer.rs [hunk 7]
```diff
@@ -717,24 +731,102 @@ impl<'a> SimpleTokenizer<'a> {
...
+    cursor: Cursor<'a>,
+}
+
+impl<'a> BackwardsTokenizer<'a> {
+    pub fn new(source: &'a str, range: TextRange, comment_range: &'a [TextRange]) -> Self {
+        Self {
+            offset: range.start(),
+            back_offset: range.end(),
+            // We could start tokenizing at a comment
+            after_newline: true,
```
[the remark this answers is not in the document]
Comment (konstin) [reply crates/ruff_python_trivia/src/tokenizer.rs:783]: this does mean we have to do the binary search when instantiating the tokenizer independent of whether we hit a newline, do you still prefer this solution? (i've also been thinking about switching)

File: crates/ruff_python_formatter/src/comments/mod.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (MichaReiser) [reply crates/ruff_python_formatter/src/comments/mod.rs:328]: On that note... we should probably move `CommentRange`s into the RC or `clone` to avoid that `clone` becomes more expensive.

File: crates/ruff_python_trivia/src/tokenizer.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (MichaReiser) [reply crates/ruff_python_trivia/src/tokenizer.rs:529]: The cursor only stores the remainder string, which is empty at this point.

File: crates/ruff_python_trivia/src/tokenizer.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (MichaReiser) [reply crates/ruff_python_trivia/src/tokenizer.rs:783]: Yeah, makes sense.

File: crates/ruff_python_formatter/src/expression/mod.rs [hunk 8]
```diff
@@ -103,9 +104,11 @@ impl FormatRule<Expr, PyFormatContext<'_>> for FormatExpr {
         });
 
         let parenthesize = match parentheses {
-            Parentheses::Preserve => {
-                is_expression_parenthesized(expression.into(), f.context().source())
-            }
+            Parentheses::Preserve => is_expression_parenthesized(
+                expression.into(),
+                f.context().comments().ranges(),
+                f.context().source(),
```
[the remark this answers is not in the document]
Comment (konstin) [reply crates/ruff_python_formatter/src/expression/mod.rs:110]: that fails because it counts as an immutable borrow of `f`

[the remark this answers is not in the document]
Comment (MichaReiser) [reply crates/ruff_python_formatter/src/expression/mod.rs:110]: Where is this a problem? Is expression parenthesized returns a bool without any lifetimes

[the remark this answers is not in the document]
Comment (konstin) [reply crates/ruff_python_formatter/src/expression/mod.rs:110]: it's not there but other call sites of `is_expression_parenthesized` or their parents which e.g. call it in closures

File: crates/ruff_python_trivia/src/tokenizer.rs (hunk 3 above)
Comment (konstin) [reply crates/ruff_python_trivia/src/tokenizer.rs:868]: ok, but it's really bad if you're doing that, the output of the tokenizer is completely wrong when you're lexing string content as rust

File: crates/ruff_python_trivia/src/tokenizer.rs [hunk 9]
```diff
@@ -717,24 +731,102 @@ impl<'a> SimpleTokenizer<'a> {
...
+///
+/// In case it finds something it can't parse, the tokenizer will return a
+/// [`SimpleTokenKind::Other`] and then only a final [`SimpleTokenKind::Bogus`] afterwards.
+pub struct BackwardsTokenizer<'a> {
+    offset: TextSize,
+    back_offset: TextSize,
+    /// Remember if we have check for comments
+    after_newline: bool,
+    /// Not `&CommentRanges` to avoid a circular dependency
+    comment_ranges: &'a [TextRange],
```
[the remark this answers is not in the document]
Comment (konstin) [reply crates/ruff_python_trivia/src/tokenizer.rs:769]: don, it's a bit odd though because the builder depends on `Tok` so it has to stay

**Comment by konstin** [comment]
Current dependencies on/for this PR:
* main
  * **PR #7360** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/7360" target="_blank">Graphite</a>  👈
    * **PR #7425** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/7425" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/astral-sh/ruff/7360?utm_source=stack-comment).
